File: zh-hant/codes/dart/chapter_greedy/fractional_knapsack.dart
/**
 * File: fractional_knapsack.dart
 * Created Time: 2023-08-11
 * Author: liuyuxin (gvenusleo@gmail.com)
 */

/* 物品 */
class Item {
  int w; // 物品重量
  int v; // 物品價值

  Item(this.w, this.v);
}

/* 分數背包：貪婪 */
double fractionalKnapsack(List<int> wgt, List<int> val, int cap) {
  // 建立物品串列，包含兩個屬性：重量、價值
  List<Item> items = List.generate(wgt.length, (i) => Item(wgt[i], val[i]));
  // 按照單位價值 item.v / item.w 從高到低進行排序
  items.sort((a, b) => (b.v / b.w).compareTo(a.v / a.w));
  // 迴圈貪婪選擇
  double res = 0;
  for (Item item in items) {
    if (item.w <= cap) {
      // 若剩餘容量充足，則將當前物品整個裝進背包
      res += item.v;
      cap -= item.w;
    } else {
      // 若剩餘容量不足，則將當前物品的一部分裝進背包
      res += item.v / item.w * cap;
      // 已無剩餘容量，因此跳出迴圈
      break;
    }
  }
  return res;
}

/* Driver Code */
void main() {
  List<int> wgt = [10, 20, 30, 40, 50];
  List<int> val = [50, 120, 150, 210, 240];
  int cap = 50;

  // 貪婪演算法
  double res = fractionalKnapsack(wgt, val, cap);
  print("不超過背包容量的最大物品價值為 $res");
}

File: ja/codes/cpp/chapter_greedy/fractional_knapsack.cpp
/**
 * File: fractional_knapsack.cpp
 * Created Time: 2023-07-20
 * Author: krahets (krahets@163.com)
 */

#include "../utils/common.hpp"

/* 品物 */
class Item {
  public:
    int w; // 品物の重さ
    int v; // 品物の価値

    Item(int w, int v) : w(w), v(v) {
    }
};

/* 分数ナップサック：貪欲法 */
double fractionalKnapsack(vector<int> &wgt, vector<int> &val, int cap) {
    // 重さと価値の 2 属性を持つ品物リストを作成
    vector<Item> items;
    for (int i = 0; i < wgt.size(); i++) {
        items.push_back(Item(wgt[i], val[i]));
    }
    // 単位価値 item.v / item.w の高い順にソートする
    sort(items.begin(), items.end(), [](Item &a, Item &b) { return (double)a.v / a.w > (double)b.v / b.w; });
    // 貪欲選択を繰り返す
    double res = 0;
    for (auto &item : items) {
        if (item.w <= cap) {
            // 残り容量が十分なら、現在の品物を丸ごとナップサックに入れる
            res += item.v;
            cap -= item.w;
        } else {
            // 残り容量が足りない場合は、現在の品物の一部だけをナップサックに入れる
            res += (double)item.v / item.w * cap;
            // 残り容量がないため、ループを抜ける
            break;
        }
    }
    return res;
}

/* Driver Code */
int main() {
    vector<int> wgt = {10, 20, 30, 40, 50};
    vector<int> val = {50, 120, 150, 210, 240};
    int cap = 50;

    // 貪欲法
    double res = fractionalKnapsack(wgt, val, cap);
    cout << "ナップサック容量を超えない最大価値は " << res << endl;

    return 0;
}
